src/session.cc

```diff
@@ -181,6 +193,150 @@ Config::debug_log(const char* fmt, ...) {
   }
 }
 
+bool
+getPublicKeyAttribute(Config& config_, CK_ATTRIBUTE_PTR pAttribute)
+{
+  switch (pAttribute->type) {
+  case CKA_ENCRYPT:
+    config_.debug_log("   Attribute: Encrypt");
+    pAttribute->ulValueLen = sizeof(CK_BBOOL);
+    if (pAttribute->pValue != nullptr) {
+      *(CK_BBOOL *)(pAttribute->pValue) = false;
+    }
+    return true;
+    break;
+
+  case CKA_VERIFY:
+    config_.debug_log("   Attribute: Verify");
+    pAttribute->ulValueLen = sizeof(CK_BBOOL);
+    if (pAttribute->pValue != nullptr) {
+      *(CK_BBOOL *)(pAttribute->pValue) = true;
+    }
+    return true;
+    break;
+
+  case CKA_VERIFY_RECOVER:
+    config_.debug_log("   Attribute: Verify Recover");
+    pAttribute->ulValueLen = sizeof(CK_BBOOL);
+    if (pAttribute->pValue != nullptr) {
+      *(CK_BBOOL *)(pAttribute->pValue) = false;
+    }
+    return true;
+    break;
+
+  case CKA_WRAP:
+    config_.debug_log("   Attribute: Wrap");
+    pAttribute->ulValueLen = sizeof(CK_BBOOL);
+    if (pAttribute->pValue != nullptr) {
+      *(CK_BBOOL *)(pAttribute->pValue) = false;
+    }
+    return true;
+    break;
+
+  case CKA_TRUSTED:
+    config_.debug_log("   Attribute: Trusted");
+    pAttribute->ulValueLen = sizeof(CK_BBOOL);
+    if (pAttribute->pValue != nullptr) {
+      *(CK_BBOOL *)(pAttribute->pValue) = true;
+    }
+    return true;
+    break;
+
+  default:
+    return false;
+  }
+}
+
+bool
+getPrivateKeyAttribute(Config& config_, CK_ATTRIBUTE_PTR pAttribute)
```

Comment:
I just made `.debug_log()` a const method, so change to const reference here. (also above)